README.md
# perpdex-oracle-contract

This repository contains the oracle smart contracts for [PerpDEX](https://perpdex.com/). For core contracts, see [perpdex-contract](https://github.com/perpdex/perpdex-contract).

## deploy

```bash
npx hardhat deploy --network rinkeby
```

## verify

```bash
npx hardhat etherscan-verify --network rinkeby --force-license --license GPL-3.0
```

## Local Development

You need Node.js 16+ to build. Use [nvm](https://github.com/nvm-sh/nvm) to install it.

Clone this repository, install Node.js dependencies, and build the source code:

```bash
git clone git@github.com:perpetual-protocol/perp-oracle-contract.git
npm i
npm run build
```

Run all the test cases:

```bash
npm run test
```

## Changelog

See [CHANGELOG](CHANGELOG.md).


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [unreleased]

-   Remove TWAP calculation from PriceFeed contracts
-   Rename `EmergencyPriceFeed` to `UniswapV3PriceFeed`

## [0.3.3] - 2022-03-18

### Changed

-   Refactor `ChainlinkPriceFeed`, `BandPriceFeed`, and `EmergencyPriceFeed` with efficient TWAP calculation.
-   Change the license to `GPL-3.0-or-later`.

## [0.3.2] - 2022-03-04

-   Using cumulative twap in Chainlink price feed

## [0.3.0] - 2022-02-07

-   Add `EmergencyPriceFeed`.

## [0.2.2] - 2021-12-28

-   `BandPriceFeed` will revert if price not updated yet.

## [0.2.1] - 2021-12-24

-   Fix `BandPriceFeed` when the price feed haven't updated.

## [0.2.0] - 2021-12-21

-   Add `BandPriceFeed`.


test/DiaPriceFeed.spec.ts
import { MockContract, smock } from "@defi-wonderland/smock"
import { expect } from "chai"
import { parseEther } from "ethers/lib/utils"
import { ethers, waffle } from "hardhat"
import { DiaPriceFeed, TestDIAOracleV2, TestDIAOracleV2__factory } from "../typechain"

interface DiaPriceFeedFixture {
    diaPriceFeed: DiaPriceFeed
    diaOracleV2: MockContract<TestDIAOracleV2>
    oracleKey: string
}

async function diaPriceFeedFixture(): Promise<DiaPriceFeedFixture> {
    const testDIAOracleV2Factory = await smock.mock<TestDIAOracleV2__factory>("TestDIAOracleV2")
    const testDIAOracleV2 = await testDIAOracleV2Factory.deploy()

    const oracleKey = "ETH/USD"
    const diaPriceFeedFactory = await ethers.getContractFactory("DiaPriceFeed")
    const diaPriceFeed = (await diaPriceFeedFactory.deploy(testDIAOracleV2.address, oracleKey)) as DiaPriceFeed

    return { diaPriceFeed, diaOracleV2: testDIAOracleV2, oracleKey }
}

describe("DiaPriceFeed Spec", () => {
    const [admin] = waffle.provider.getWallets()
    const loadFixture: ReturnType<typeof waffle.createFixtureLoader> = waffle.createFixtureLoader([admin])
    let diaPriceFeed: DiaPriceFeed
    let diaOracleV2: MockContract<TestDIAOracleV2>
    let currentTime: number
    let roundData: any[]

    async function updatePrice(price: number, forward: boolean = true): Promise<void> {
        roundData.push([parseEther(price.toString()), currentTime])
        diaOracleV2.getValue.returns(() => {
            return roundData[roundData.length - 1]
        })

        if (forward) {
            currentTime += 15
            await ethers.provider.send("evm_setNextBlockTimestamp", [currentTime])
            await ethers.provider.send("evm_mine", [])
        }
    }

    async function setCurrentTimeToLatest() {
        const latestTimestamp = (await waffle.provider.getBlock("latest")).timestamp
        currentTime = latestTimestamp
    }

    beforeEach(async () => {
        const _fixture = await loadFixture(diaPriceFeedFixture)
        diaOracleV2 = _fixture.diaOracleV2
        diaPriceFeed = _fixture.diaPriceFeed
        roundData = []
    })

    describe("decimals", () => {
        it("return 8", async () => {
            expect(await diaPriceFeed.decimals()).to.be.eq(8)
        })
    })

    describe("getPrice", () => {
        describe("when price is not updated yet", () => {
            it("revert", async () => {
                await expect(diaPriceFeed.getPrice()).to.be.revertedWith("DPF_GOD: time is zero")
            })
        })

        describe("when price <= 0 ", () => {
            beforeEach(async () => {
                await setCurrentTimeToLatest()
                await updatePrice(0, false)
            })
            it("revert", async () => {
                await expect(diaPriceFeed.getPrice()).to.be.revertedWith("DPF_GOD: invalid price")
            })
        })

        describe("when price is updated", () => {
            beforeEach(async () => {
                await setCurrentTimeToLatest()
                await updatePrice(400, false)
            })
            it("return spot price", async () => {
                expect(await diaPriceFeed.getPrice()).to.eq(parseEther("400"))
            })
        })
    })
})


test/BandPriceFeed.spec.ts
import { MockContract, smock } from "@defi-wonderland/smock"
import { expect } from "chai"
import { parseEther } from "ethers/lib/utils"
import { ethers, waffle } from "hardhat"
import { BandPriceFeed, TestStdReference, TestStdReference__factory } from "../typechain"

interface BandPriceFeedFixture {
    bandPriceFeed: BandPriceFeed
    bandReference: MockContract<TestStdReference>
    baseAsset: string
}

async function bandPriceFeedFixture(): Promise<BandPriceFeedFixture> {
    const testStdReferenceFactory = await smock.mock<TestStdReference__factory>("TestStdReference")
    const testStdReference = await testStdReferenceFactory.deploy()

    const baseAsset = "ETH"
    const bandPriceFeedFactory = await ethers.getContractFactory("BandPriceFeed")
    const bandPriceFeed = (await bandPriceFeedFactory.deploy(testStdReference.address, baseAsset)) as BandPriceFeed

    return { bandPriceFeed, bandReference: testStdReference, baseAsset }
}

describe("BandPriceFeed/CumulativeTwap Spec", () => {
    const [admin] = waffle.provider.getWallets()
    const loadFixture: ReturnType<typeof waffle.createFixtureLoader> = waffle.createFixtureLoader([admin])
    let bandPriceFeed: BandPriceFeed
    let bandReference: MockContract<TestStdReference>
    let currentTime: number
    let roundData: any[]

    async function updatePrice(price: number, forward: boolean = true): Promise<void> {
        roundData.push([parseEther(price.toString()), currentTime, currentTime])
        bandReference.getReferenceData.returns(() => {
            return roundData[roundData.length - 1]
        })

        if (forward) {
            currentTime += 15
            await ethers.provider.send("evm_setNextBlockTimestamp", [currentTime])
            await ethers.provider.send("evm_mine", [])
        }
    }

    beforeEach(async () => {
        const _fixture = await loadFixture(bandPriceFeedFixture)
        bandReference = _fixture.bandReference
        bandPriceFeed = _fixture.bandPriceFeed
        roundData = []
    })

    describe("getPrice", () => {
        beforeEach(async () => {
            // `base` = now - _interval
            // bandReference's answer
            // timestamp(base + 0)  : 400
            // timestamp(base + 15) : 405
            // timestamp(base + 30) : 410
            // now = base + 45
            //
            //  --+------+-----+-----+-----+-----+-----+
            //          base                          now
            const latestTimestamp = (await waffle.provider.getBlock("latest")).timestamp
            currentTime = latestTimestamp

            await updatePrice(400)
            await updatePrice(405)
            await updatePrice(410)
        })

        it("return latest price", async () => {
            const price = await bandPriceFeed.getPrice()
            expect(price).to.eq(parseEther("410"))
        })
    })
})


test/PriceFeed.gas.test.ts
import { parseEther } from "ethers/lib/utils"
import { ethers, waffle } from "hardhat"
import { BandPriceFeed, ChainlinkPriceFeed, TestAggregatorV3, TestPriceFeed, TestStdReference } from "../typechain"

interface PriceFeedFixture {
    bandPriceFeed: BandPriceFeed
    bandReference: TestStdReference
    baseAsset: string

    // chainlinik
    chainlinkPriceFeed: ChainlinkPriceFeed
    aggregator: TestAggregatorV3
}

async function priceFeedFixture(): Promise<PriceFeedFixture> {
    // band protocol
    const testStdReferenceFactory = await ethers.getContractFactory("TestStdReference")
    const testStdReference = await testStdReferenceFactory.deploy()

    const baseAsset = "ETH"
    const bandPriceFeedFactory = await ethers.getContractFactory("BandPriceFeed")
    const bandPriceFeed = (await bandPriceFeedFactory.deploy(testStdReference.address, baseAsset)) as BandPriceFeed

    // chainlink
    const testAggregatorFactory = await ethers.getContractFactory("TestAggregatorV3")
    const testAggregator = await testAggregatorFactory.deploy()

    const chainlinkPriceFeedFactory = await ethers.getContractFactory("ChainlinkPriceFeed")
    const chainlinkPriceFeed = (await chainlinkPriceFeedFactory.deploy(testAggregator.address)) as ChainlinkPriceFeed

    return { bandPriceFeed, bandReference: testStdReference, baseAsset, chainlinkPriceFeed, aggregator: testAggregator }
}

describe.skip("Price feed gas test", () => {
    const [admin] = waffle.provider.getWallets()
    const loadFixture: ReturnType<typeof waffle.createFixtureLoader> = waffle.createFixtureLoader([admin])
    let bandPriceFeed: BandPriceFeed
    let bandReference: TestStdReference
    let chainlinkPriceFeed: ChainlinkPriceFeed
    let aggregator: TestAggregatorV3
    let currentTime: number
    let testPriceFeed: TestPriceFeed
    let beginPrice = 400
    let round: number

    async function updatePrice(price: number, forward: boolean = true): Promise<void> {
        await bandReference.setReferenceData({
            rate: parseEther(price.toString()),
            lastUpdatedBase: currentTime,
            lastUpdatedQuote: currentTime,
        })

        await aggregator.setRoundData(round, parseEther(price.toString()), currentTime, currentTime, round)

        if (forward) {
            currentTime += 15
            await ethers.provider.send("evm_setNextBlockTimestamp", [currentTime])
            await ethers.provider.send("evm_mine", [])
        }
    }

    before(async () => {
        const _fixture = await loadFixture(priceFeedFixture)
        bandReference = _fixture.bandReference
        bandPriceFeed = _fixture.bandPriceFeed
        chainlinkPriceFeed = _fixture.chainlinkPriceFeed
        aggregator = _fixture.aggregator
        round = 0

        const TestPriceFeedFactory = await ethers.getContractFactory("TestPriceFeed")
        testPriceFeed = (await TestPriceFeedFactory.deploy(
            chainlinkPriceFeed.address,
            bandPriceFeed.address,
        )) as TestPriceFeed

        currentTime = (await waffle.provider.getBlock("latest")).timestamp
        for (let i = 0; i < 255; i++) {
            round = i
            await updatePrice(beginPrice + i)
        }
    })

    it("band protocol ", async () => {
        await testPriceFeed.fetchBandProtocolPrice()
    })

    it("chainlink", async () => {
        await testPriceFeed.fetchChainlinkPrice()
    })
})


hardhat.config.ts
import "@nomiclabs/hardhat-ethers"
import "@nomiclabs/hardhat-waffle"
import "@typechain/hardhat"
import "hardhat-dependency-compiler"
import "hardhat-deploy"
import "hardhat-gas-reporter"
import { HardhatUserConfig } from "hardhat/config"
import { config as dotenvConfig } from "dotenv"
import "solidity-coverage"
import { resolve } from "path"
import "@matterlabs/hardhat-zksync-solc"

dotenvConfig({ path: resolve(__dirname, "./.env") })

const config: HardhatUserConfig = {
    solidity: {
        version: "0.8.12",
        settings: {
            optimizer: { enabled: true, runs: 200 },
            evmVersion: "berlin",
            // for smock to mock contracts
            outputSelection: {
                "*": {
                    "*": ["storageLayout"],
                },
            },
        },
    },
    zksolc: {
        version: "0.1.0",
        compilerSource: "docker",
        settings: {
            optimizer: {
                enabled: true,
            },
            experimental: {
                dockerImage: "matterlabs/zksolc",
            },
        },
    },
    namedAccounts: {
        deployer: 0,
    },
    networks: {
        hardhat: {
            allowUnlimitedContractSize: true,
        },
    },
    gasReporter: {
        enabled: true,
    },
}

if (process.env.TESTNET_PRIVATE_KEY) {
    if (process.env.INFURA_PROJECT_ID) {
        config.networks.rinkeby = {
            url: "https://rinkeby.infura.io/v3/" + process.env.INFURA_PROJECT_ID,
            accounts: [process.env.TESTNET_PRIVATE_KEY],
            gasMultiplier: 2,
        }
    }

    config.networks.mumbai = {
        url: "https://rpc-mumbai.maticvigil.com",
        accounts: [process.env.TESTNET_PRIVATE_KEY],
        gasMultiplier: 2,
    }

    config.networks.fuji = {
        url: "https://api.avax-test.network/ext/bc/C/rpc",
        accounts: [process.env.TESTNET_PRIVATE_KEY],
        gasMultiplier: 2,
    }

    config.networks.shibuya = {
        url: "https://evm.shibuya.astar.network",
        chainId: 81,
        accounts: [process.env.TESTNET_PRIVATE_KEY],
        gasMultiplier: 2,
        verify: {
            etherscan: {
                apiUrl: "https://blockscout.com/shibuya",
            },
        },
    }

    config.networks.zksync2_testnet = {
        url: "https://zksync2-testnet.zksync.dev",
        chainId: 280,
        accounts: [process.env.TESTNET_PRIVATE_KEY],
        gasMultiplier: 2,
        verify: {
            etherscan: {
                apiUrl: "https://zksync2-testnet.zkscan.io/",
            },
        },
        zksync: true,
    }
}

export default config


scripts/univ2_survey.ts
import { ethers } from "hardhat"
import IUniswapV2Pair from "@uniswap/v2-core/build/IUniswapV2Pair.json"
import IUniswapV2Factory from "@uniswap/v2-core/build/IUniswapV2Factory.json"
import IUniswapV2Router02 from "@uniswap/v2-periphery/build/IUniswapV2Router02.json"

async function main() {
    // zksync testnet
    const routerAddress = "0x82Ec84c7368bb9089E1077c6e1703675c35A4237"
    const usdcAddress = "0x54a14D7559BAF2C8e8Fa504E019d32479739018c"
    const wethAddress = "0xB4fbFB7807C31268Dc1ac8c26fA4ef41115d0ece"

    const [signer] = await ethers.getSigners()
    const router = new ethers.Contract(routerAddress, IUniswapV2Router02.abi, signer)
    const factoryAddress = await router.factory()
    console.log("factory address: " + factoryAddress)

    const factory = new ethers.Contract(factoryAddress, IUniswapV2Factory.abi, signer)
    const pairAddress = await factory.getPair(wethAddress, usdcAddress)
    console.log("pair address: " + pairAddress)

    const pair = new ethers.Contract(pairAddress, IUniswapV2Pair.abi, signer)
    console.log("token0 address: " + (await pair.token0()))
    console.log("token1 address: " + (await pair.token1()))
}

main()
    .then(() => process.exit(0))
    .catch(error => {
        console.error(error)
        process.exit(1)
    })


deploy/01_deploy_chainlink.ts
import { HardhatRuntimeEnvironment } from "hardhat/types"
import { DeployFunction } from "hardhat-deploy/types"
import { parseEther } from "ethers/lib/utils"

const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
    const { deployments, getNamedAccounts, getChainId } = hre
    const { deploy, execute } = deployments
    const { deployer } = await getNamedAccounts()

    const chainlinkAddress = {
        "31337": "0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419", // ETH(mainnet) hardhat
        "4": "0x8A753747A1Fa494EC906cE90E9f37563A8AF630e", // ETH rinkeby
        "80001": "0x0715A7794a1dc8e42615F059dD6e406A6594651A", // ETH mumbai
        "43113": "0x0A77230d17318075983913bC2145DB16C7366156", // AVAX fuji
    }[await getChainId()]

    const chainlinkSymbol = {
        "31337": 'ETHUSD', // ETH(mainnet) hardhat
        "4": "ETHUSD", // ETH rinkeby
        "80001": 'ETHUSD', // ETH mumbai
        "43113": "AVAXUSD", // AVAX fuji
    }[await getChainId()]

    await deploy("ChainlinkPriceFeed" + chainlinkSymbol, {
        contract: 'ChainlinkPriceFeed',
        from: deployer,
        args: [chainlinkAddress],
        log: true,
        autoMine: true,
    })
}

export default func
func.tags = ["ChainlinkPriceFeed"]


deploy/01_deploy_price_feed.ts
import { HardhatRuntimeEnvironment } from "hardhat/types"
import { DeployFunction } from "hardhat-deploy/types"
import { parseEther } from "ethers/lib/utils"

const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
    const { deployments, getNamedAccounts, getChainId } = hre
    const { deploy, execute } = deployments
    const { deployer } = await getNamedAccounts()

    const configs = {
        hardhat: [
            // hardhat (ETH mainnet)
            {
                contract: "ChainlinkPriceFeed",
                symbol: "ETHUSD",
                args: ["0x5f4eC3Df9cbd43714FE2740f5E3616155c5b8419"],
            },
        ],
        rinkeby: [
            // ETH rinkeby
            {
                contract: "ChainlinkPriceFeed",
                symbol: "ETHUSD",
                args: ["0x8A753747A1Fa494EC906cE90E9f37563A8AF630e"],
            },
            {
                contract: "ChainlinkPriceFeed",
                symbol: "BTCUSD",
                args: ["0xECe365B379E1dD183B20fc5f022230C044d51404"],
            },
            {
                contract: "ChainlinkPriceFeed",
                symbol: "MATICUSD",
                args: ["0x7794ee502922e2b723432DDD852B3C30A911F021"],
            },
            {
                contract: "ChainlinkPriceFeed",
                symbol: "LINKUSD",
                args: ["0xd8bD0a1cB028a31AA859A21A3758685a95dE4623"],
            },
        ],
        shibuya: [
            // astar shibuya https://docs.diadata.org/documentation/oracle-documentation/deployed-contracts
            {
                contract: "DiaPriceFeed",
                symbol: "BTCUSD",
                args: ["0x1232AcD632Dd75f874E357c77295Da3f5Cd7733E", "BTC/USD"],
            },
            {
                contract: "DiaPriceFeed",
                symbol: "ETHUSD",
                args: ["0x1232AcD632Dd75f874E357c77295Da3f5Cd7733E", "ETH/USD"],
            },
            {
                contract: "DiaPriceFeed",
                symbol: "KSMUSD",
                args: ["0x1232AcD632Dd75f874E357c77295Da3f5Cd7733E", "KSM/USD"],
            },
            {
                contract: "DiaPriceFeed",
                symbol: "SDNUSD",
                args: ["0x1232AcD632Dd75f874E357c77295Da3f5Cd7733E", "SDN/USD"],
            },
            {
                contract: "DiaPriceFeed",
                symbol: "DIAUSD",
                args: ["0x1232AcD632Dd75f874E357c77295Da3f5Cd7733E", "DIA/USD"],
            },
            {
                contract: "DiaPriceFeed",
                symbol: "ASTRUSD",
                args: ["0x1232AcD632Dd75f874E357c77295Da3f5Cd7733E", "ASTR/USD"],
            },
        ],
        mumbai: [
            // ETH mumbai
            {
                contract: "ChainlinkPriceFeed",
                symbol: "ETHUSD",
                args: ["0x0715A7794a1dc8e42615F059dD6e406A6594651A"],
            },
        ],
        fuji: [
            // AVAX fuji
            {
                contract: "ChainlinkPriceFeed",
                symbol: "AVAXUSD",
                args: ["0x0A77230d17318075983913bC2145DB16C7366156"],
            },
        ],
        zksync2_testnet: [
            // use scripts/univ2_survey.ts to get addresses
            {
                contract: "UniswapV2PriceFeed",
                symbol: "ETHUSD",
                args: ["0xE9ea01488213269fDE4260B09e60faC36C971AD5", true],
            },
        ],
    }[hre.network.name]

    for (let i = 0; i < configs.length; i++) {
        const config = configs[i]
        await deploy(config.contract + config.symbol, {
            contract: config.contract,
            from: deployer,
            args: config.args,
            log: true,
            autoMine: true,
        })
    }
}

export default func
func.tags = ["price_feed"]
